File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
  def generateParenthesis(self, available: int, s: str = '', opened: int = 0) -> List[str]:
    if opened and available:
      return self.generateParenthesis(available - 1, s + '(', opened + 1) + \
             self.generateParenthesis(available, s + ')', opened - 1)
    elif opened:
      return self.generateParenthesis(available, s + ')', opened - 1)
    elif available:
      return self.generateParenthesis(available - 1, s + '(', opened + 1)
    return [s]
        

File: 0022-generate-parentheses/0022-generate-parentheses.php
class Solution {

  /**
    * @param Integer $n
    * @return String[]
    */
  function generateParenthesis($available, $s = '', $opened = 0) {
    if ($opened && $available) {
      return array_merge(
          $this->generateParenthesis($available - 1, $s . '(', $opened + 1),
          $this->generateParenthesis($available, $s . ')', $opened - 1),
      );
    } else if ($opened) {
      return $this->generateParenthesis($available, $s . ')', $opened - 1);
    } else if ($available) {
      return $this->generateParenthesis($available - 1, $s . '(', $opened + 1);
    }
    return [ $s ];
  }
}